homeassistant/helpers/service.py

```diff
@@ -352,3 +356,40 @@ def async_register_admin_service(
     hass.services.async_register(
         domain, service, admin_handler, schema
     )
+
+
+@bind_hass
+def verify_domain_control(hass: HomeAssistantType, domain: str) -> Callable:
+    """Ensure permission to access any entity under domain in service call."""
+    def decorator(service: 'Service') -> Callable:
+        """Decorate."""
+        @wraps(service)
+        async def check_permissions(
+                call: 'ServiceCall') -> Any:
+            """Check user permission and raise before call if unauthorized."""
+            if not call.context.user_id:
+                return await service(call)
+
+            user = await hass.auth.async_get_user(call.context.user_id)
+            if user is None:
+                raise UnknownUser(
+                    context=call.context, permission=POLICY_CONTROL)
+
+            reg = await hass.helpers.entity_registry.async_get_registry()
+            entities = [
+                entity.entity_id for entity in reg.entities.values()
+                if entity.platform == domain
+            ]
+
+            for entity_id in entities:
+                if user.permissions.check_entity(entity_id, POLICY_CONTROL):
```

Comment:
This block missed in tests coverage that means you don't have a test to cover positive case.

Also, is it more descriptive if we provide a helper method to wrap L378-L385 such as `hass.helpers.permissons.check_domain(user, domain)`

Reply:
Or maybe `user.permissions.check_entity(domain + '.*', POLICY_CONTROL)`? Haven't touch permission for a while, need reread the code to fresh my memory.

Reply:
I'm not sure where else this new helper would be used – seems outside the scope of this PR's MVP, no?

Reply:
After a quick review the implementation, I indeed think we should provide a function in permissions itself, not in the helpers

`check_domain()`

Reply:
I am not a fan of premature generalizing it without making sure we need a generic function.

Reply:
@balloob Would you agree with @awarecan about moving this into permissions itself (instead of permissions helpers)?

Reply:
I would hesitate moving it there unless we have more than 1 use case.